# 가설검정 단계

In [2]:
import warnings
warnings.filterwarnings('ignore')

from matplotlib import rcParams
rcParams['font.family'] = 'Gulim'
rcParams['axes.unicode_minus'] = False

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats as stats
%matplotlib inline
plt.rc('figure', figsize=(10, 6))
from matplotlib import rcParams
rcParams['font.family'] = 'New Gulim'
rcParams['font.size'] = 10
rcParams['axes.unicode_minus'] = False

대한민국 남성의 평균 몸무게는 70 킬로그램이다

In [ ]:
#1. 가설수립
# 귀무: 대한민국 남성의 평균 몸무게는 70 킬로그램이다
# 연구: 대한민국 남성의 평균 몸무게는 70 킬로그램이다

In [9]:
# 2. 통계량 산출 - 데이터 수집
np.random.seed(123)
weight = np.random.uniform(40, 100, 100)

test_result = stats.ttest_1samp(weight, 70)
np.mean(weight)
print(f'평균 : {np.mean(weight)}')
print(f'p-value : {test_result.pvalue}')
# weight 
# 3. 기각/채택
print(f' 검정결과 : {'귀무가설 기각' if test_result.pvalue < 0.05 else '귀무가설 성립'} ')
# 4. 결정
if test_result.pvalue < 0.05:
    print('대한민국 남성의 평균 몸무게는 70킬로가 아니다')
else:
    print('대한민국 남성의 평균 몸무게는 70킬로이다')

평균 : 70.08573280308887
p-value : 0.953913826658112
 검정결과 : 귀무가설 성립 
대한민국 남성의 평균 몸무게는 70킬로이다


In [4]:
# 표본 데이터 (30명)
coffee_data = [2.9, 2.4, 2.7, 3.0, 2.6, 2.8, 2.9, 3.1, 2.7, 2.6,
               2.8, 2.5, 3.2, 3.0, 2.4, 2.6, 2.7, 3.1, 2.8, 2.7,
               2.5, 2.9, 2.6, 2.8, 2.9, 2.5, 2.7, 2.8, 3.0, 2.9]

# 단일표본 t-검정
t_stat, p_value = stats.ttest_1samp(coffee_data, popmean=2.5)

print(f"t값 = {t_stat:.3f}, p값 = {p_value:.3f}")

print(f' 검정결과 : {'귀무가설 기각' if p_value < 0.05 else '귀무가설 성립'} ')
# 4. 결정
if p_value < 0.05:
    print('설문조사와 실제 커피마신 잔의 수는 차이가 있다')
else:
    print('설문조사와 실제 커피마신 잔의 수는 차이가 없다')

t값 = 7.031, p값 = 0.000
 검정결과 : 귀무가설 기각 
설문조사와 실제 커피마신 잔의 수는 차이가 있다


In [5]:
group_A = [78, 85, 82, 88, 76, 81, 79, 84, 82, 80]
group_B = [72, 75, 78, 74, 71, 77, 70, 76, 75, 73]

t_stat, p_value = stats.ttest_ind(group_A, group_B, equal_var=False)
print(f"t값 = {t_stat:.3f}, p값 = {p_value:.3f}")

t값 = 5.331, p값 = 0.000


H0: 두 그룹의 평균은 차이가 없다 = 같다     
H1: 두 그룹의 평균은 차이가 있다

통계량: 5.33, 표준오차의 5배가 차이난다      
유의확률: 0.000... < 0.05

귀무가설 기각/채택 여부: 귀무가설 기각  
결론: 두 그룹의 평균은 유의미하게 차이가 있다

In [ ]:
# 대응 표본 t-검정 예제: 약물 투여 전후 혈압 비교
before = [120, 122, 143, 130, 135, 118, 127, 140, 132, 125]
after = [115, 117, 138, 125, 129, 112, 120, 135, 126, 119]

t_stat, p_value = stats.ttest_rel(before, after)
# rel과 ind의 차이점은 같은 집단의 전후냐 서로 다른 두 집단이냐의 차이
# 즉 rel은 두 집다닝 서로 의존성을 띄며 ind와 다르게 쌍으로 바교를 한다
# rel은 각 차이들의 평균 / ind는 전체평균을 비교
print(f"t값 = {t_stat:.3f}, p값 = {p_value:.5f}")

t값 = 25.327, p값 = 0.00000


H0: 두 그룹의 혈압은 차이가 없다 = 같다     
H1: 두 그룹의 혈압은 차이가 있다    

통계량: 25.327, 표준오차의 약 25배가 차이난다      
유의확률: 0.00 < 0.05

귀무가설 기각/채택 여부: 귀무가설 기각      
결론: 두 그룹의 혈압은 유의미하게 차이가 있다

In [11]:
import numpy as np
obs = np.array([[40, 10],  # 남자 (좋아함, 안좋아함) 교차표
                [30, 20]]) # 여자

chi2, p, dof, expected = stats.chi2_contingency(obs)
print(f"chi2 = {chi2:.3f}, p값 = {p:.5f}")

chi2 = 3.857, p값 = 0.04953


H0: 두 그룹의 평균은 차이가 없다 = 같다     
H1: 두 그룹의 평균은 차이가 있다

통계량: 3.857, 표준오차의 약 4배가 차이난다      
유의확률: 0.049 < 0.05

귀무가설 기각/채택 여부: 귀무가설 기각 
결론: 두 그룹의 평균은 유의미하게 차이가 있다

In [12]:
from scipy.stats import f_oneway

# 세 그룹의 점수
group1 = [80, 85, 90]
group2 = [70, 75, 78]
group3 = [60, 65, 62]

f_stat, p_value = f_oneway(group1, group2, group3)
print(f"F값: {f_stat:.2f}, p값: {p_value:.4f}")

F값: 24.28, p값: 0.0013


H0: 세 그룹의 평균은 차이가 없다 = 같다     
H1: 적어도 1개의 그룹은 다르다

결과: f = 24.3, 그룹 간 분산의 차이가 그룹 내 분산의 차이보다 크다  
유의확률: 0.013 < 0.05

귀무가설 기각/채택 여부: 귀무가설 기각 
결론: 두 그룹의 평균은 유의미하게 차이가 있다